### PR TITLE
ARROW-11673 - [C++] Casting dictionary type to use different index type

### PR DESCRIPTION
**Pull request by nirandaperera**
This PR adds casting from one dictionary type to anther dictionary type for both scalars and arrays :
ex:
```
dictionary(int8(), int16()) --> dictionary(int32(), int64())
```

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        ArrayData::Make(in_type->index_type(), in_array->length, in_array->buffers,
+                        in_array->GetNullCount(), in_array->offset);
+    ARROW_ASSIGN_OR_RAISE(auto casted_indices, Cast(indices_arr, out_type->index_type(),
+                                                    options, ctx->exec_context()));
+    out_array->buffers[0] = std::move(casted_indices.array()->buffers[0]);
+    out_array->buffers[1] = std::move(casted_indices.array()->buffers[1]);
+  }
+
+  // data (dict)
+  if (in_type->index_type()->Equals(out_type->index_type())) {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:97]: Shouldn't this be checking value_type?

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+  // if input is array
+  const std::shared_ptr<ArrayData>& in_array = batch[0].array();
+  auto in_type = std::static_pointer_cast<DictionaryType>(in_array->type);
+
+  ArrayData* out_array = out->mutable_array();
+
+  if (in_type->index_type()->Equals(out_type->index_type())) {
+    out_array->buffers[0] = in_array->buffers[0];
+    out_array->buffers[1] = in_array->buffers[1];
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:84]: What if there's an offset? I suppose it does pass CheckCast which does do slicing…

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:84]: It would also be nice to copy null_count.

File: cpp/src/arrow/CMakeLists.txt [hunk 3]
```diff
@@ -385,7 +385,8 @@ if(ARROW_COMPUTE)
        compute/kernels/scalar_arithmetic.cc
        compute/kernels/scalar_boolean.cc
        compute/kernels/scalar_cast_boolean.cc
-       compute/kernels/scalar_cast_internal.cc
+          compute/kernels/scalar_cast_dictionary.cc
+          compute/kernels/scalar_cast_internal.cc
```
Comment (lidavidm) [inline cpp/src/arrow/CMakeLists.txt:389]: nit: this needs to be cmake-formatted

File: cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc (hunk 1 above)
Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_cast_dictionary.cc:97]: Lol! indeed! :open_mouth:

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 4]
```diff
@@ -1931,14 +1931,29 @@ TEST(Cast, DictTypeToAnotherDict) {
...
   EXPECT_RAISES_WITH_MESSAGE_THAT(
       Invalid, testing::HasSubstr("Integer value 1000 not in range"),
       Cast(arr, dictionary(int8(), int8()), CastOptions::Safe()));
+
+  // check unsafe indices
+  random::RandomArrayGenerator rand(/*seed=*/0);
+  int64_t len = 1000;
+  auto val_arr = rand.ArrayOf(int32(), len, /*null_probability=*/0.01);
+  ASSERT_OK_AND_ASSIGN(auto arr2, DictionaryEncode(val_arr));
+  // check unsafe indices. Cannot validate this array because ValidateOutput throws an
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: Sorry - now that I look at the JIRA again, it seems having unsafe casts isn't useful for this case? What do you think? If it produces invalid output (presumably, negative and out of bounds indices?), then it seems kind of pointless.

Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: I was also thinking about this. OTOH I was thinking whether the ValidataOutput should check the validity of dictionary array in DictionaryType. :thinking:

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: It looks like it does - are you saying it shouldn't?

**Comment by nirandaperera** [comment]
@lidavidm is this a known test failure? `[  FAILED  ] Expression.ReplaceFieldsWithKnownValues`
https://github.com/apache/arrow/pull/10721/checks?check_run_id=3078144913

**Comment by lidavidm** [comment]
```
Expected 'ReplaceFieldsWithKnownValues( KnownFieldValues{{{"dict_str", dict_i32}}}, expr)' to fail with NotImplemented, but got OK
```

Looks like this enables something that didn't work before?

```
  ASSERT_OK_AND_ASSIGN(auto expr, field_ref("dict_str").Bind(*kBoringSchema));
  Datum dict_i32{
      DictionaryScalar::Make(MakeScalar<int32_t>(0), ArrayFromJSON(int32(), R"([3])"))};
  // Unsupported cast dictionary(int32(), int32()) -> dictionary(int32(), utf8())
  ASSERT_RAISES(NotImplemented, ReplaceFieldsWithKnownValues(
                                    KnownFieldValues{{{"dict_str", dict_i32}}}, expr));
  // Unsupported cast dictionary(int8(), utf8()) -> dictionary(int32(), utf8())
  dict_str = Datum{
      DictionaryScalar::Make(MakeScalar<int8_t>(0), ArrayFromJSON(utf8(), R"(["a"])"))};
  ASSERT_RAISES(NotImplemented, ReplaceFieldsWithKnownValues(
                                    KnownFieldValues{{{"dict_str", dict_str}}}, expr));
```

**Comment by nirandaperera** [comment]
> ```
> Expected 'ReplaceFieldsWithKnownValues( KnownFieldValues{{{"dict_str", dict_i32}}}, expr)' to fail with NotImplemented, but got OK
> ```
> 
> Looks like this enables something that didn't work before?
> 
> ```
>   ASSERT_OK_AND_ASSIGN(auto expr, field_ref("dict_str").Bind(*kBoringSchema));
>   Datum dict_i32{
>       DictionaryScalar::Make(MakeScalar<int32_t>(0), ArrayFromJSON(int32(), R"([3])"))};
>   // Unsupported cast dictionary(int32(), int32()) -> dictionary(int32(), utf8())
>   ASSERT_RAISES(NotImplemented, ReplaceFieldsWithKnownValues(
>                                     KnownFieldValues{{{"dict_str", dict_i32}}}, expr));
>   // Unsupported cast dictionary(int8(), utf8()) -> dictionary(int32(), utf8())
>   dict_str = Datum{
>       DictionaryScalar::Make(MakeScalar<int8_t>(0), ArrayFromJSON(utf8(), R"(["a"])"))};
>   ASSERT_RAISES(NotImplemented, ReplaceFieldsWithKnownValues(
>                                     KnownFieldValues{{{"dict_str", dict_str}}}, expr));
> ```

Ah! indeed! thanks...

File: cpp/src/arrow/compute/exec/expression_test.cc [hunk 5]
```diff
@@ -942,14 +942,14 @@ TEST(Expression, ReplaceFieldsWithKnownValues) {
   ASSERT_OK_AND_ASSIGN(auto expr, field_ref("dict_str").Bind(*kBoringSchema));
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/expression_test.cc:942]: nit: this isn't used anymore

**Comment by lidavidm** [comment]
I think this is good now, thanks! (Just one nit about an unused variable)

**Comment by lidavidm** [comment]
I take that back, it's used after all.

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 4 above)
Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: I dont think ValidateOutput checks the validity of `ArrayData::dictionary` field. That's why ValidateOutput passes for unsafe valies AFAIU. :thin

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: It looks like it does: https://github.com/apache/arrow/blob/dbeed527dff79f3fa4576722a821fb7af3abd878/cpp/src/arrow/array/validate.cc#L565-L572

Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: I think there's no way a primitive array to be validated once the casting is complete (unsafely in this instance).

**Review by lidavidm — APPROVED** [review]
Thanks for taking this.

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 4 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: We're casting from dictionary to dictionary here though. 

Anyways, the point stands: this is an unsafe cast that generates an invalid array, so is this case worth supporting?

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 6]
```diff
@@ -1911,5 +1911,50 @@ TEST(Cast, ExtensionTypeToIntDowncast) {
...
+  };
+
+  //    check same type passed on to casting
+  check_cast(dictionary(int8(), int16()), dictionary(int8(), int16()),
+             "[1, 2, 3, 1, null, 3]");
+  check_cast(dictionary(int8(), int16()), dictionary(int32(), int64()),
+             "[1, 2, 3, 1, null, 3]");
+  check_cast(dictionary(int32(), utf8()), dictionary(int8(), utf8()),
+             R"(["a", "b", "a", null])");
+
```
Comment (edponce) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1932]: Add tests casting from int to/from float and signed to/from unsigned integers.

**Review by edponce — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 6 above)
Comment (edponce) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1932]: I tested
```
  check_cast(dictionary(int32(), int32()), dictionary(int8(), float32()),
             "[1, 2, 3, 1, null, 3]");
```
and get the following error:
`NotImplemented: JSON conversion to dictionary<values=float, indices=int8, ordered=0> not implemented`

Comment (edponce) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1932]: I ran int to/from float and unsigned (with Safe and Unsafe options) and it works correctly. LGTM.

**Review by edponce — APPROVED** [review]

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 4 above)
Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1948]: I think I agree with you. It's not worth testing that. I will remove it. :-)

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 7]
```diff
@@ -1930,29 +1930,23 @@ TEST(Cast, DictTypeToAnotherDict) {
   check_cast(dictionary(int32(), utf8()), dictionary(int8(), utf8()),
              R"(["a", "b", "a", null])");
 
+  // check float types (NOTE: ArrayFromJSON doesnt work for float value dictionary types)
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1933]: What was the error here? Would be nice to file a JIRA for it

**Review by lidavidm — APPROVED** [review]

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 7 above)
Comment (nirandaperera) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: Done!

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 8]
```diff
@@ -1930,7 +1930,7 @@ TEST(Cast, DictTypeToAnotherDict) {
   check_cast(dictionary(int32(), utf8()), dictionary(int8(), utf8()),
              R"(["a", "b", "a", null])");
 
-  // check float types (NOTE: ArrayFromJSON doesnt work for float value dictionary types)
+  // check float types (TODO: ARROW-13381 ArrayFromJSON doesnt work for float value dictionary types)
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1933]:
```suggestion
  // check float types
  // TODO(ARROW-13381): ArrayFromJSON doesnt work for float value dictionary types
```

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1933]: Needs to be linted.

**Comment by nirandaperera** [comment]
I made the suggested changes and I think this is ready now

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 7 above)
Comment (edponce) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: The error is from `ArrayFromJSON(dictionary(..., floatXX()), ...)` is:
```
NotImplemented: JSON conversion to dictionary<values=float, indices=int8, ordered=0> not implemented
```
Nevertheless, you can `Cast` successfully to `float` values with
```
auto arr = ArrayFromJSON(dictionary(int8(), int32()), "[1, 2, 3, 1, null, 3]");
ASSERT_OK_AND_ASSIGN(auto casted, Cast(arr, dictionary(int8(), float32()), CastOptions::Safe()));
```

Comment (edponce) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: @nirandaperera I submitted [this PR](https://github.com/apache/arrow/pull/10746) that enables using `ArrayFromJSON` for dictionaries with floating-point values. If PR passes and gets merged, you can change test accordingly and remove error comment.

**Review by kszucs — APPROVED** [review]
Thanks Niranda! Merging on green.

**Pull request closed without merging**